### DISCUSSION TITLE
How to implement recurrent dropout with NNX ?

### DISCUSSION THREAD

**Discussion by sgkouzias** [google/flax#4397] [Q&A] [answered]
I am trying to replicate the implementation below from the book "_Deep Learning with Python_" by François Chollet regarding [timeseries data](https://s3.amazonaws.com/keras-datasets/jena_climate_2009_2016.csv.zip):

```bash
import keras
from keras import layers

sequence_length = 120
num_features = 14

inputs = keras.Input(shape=(sequence_length, num_features))
x = layers.LSTM(32, recurrent_dropout=0.25)(inputs)
x = layers.Dropout(0.5)(x)
outputs = layers.Dense(1)(x)
model = keras.Model(inputs, outputs)
```

My attempt below:

```bash
class RecurrentDropout(nnx.Module):
    """Recurrent Dropout Layer.

    This module implements recurrent dropout, a regularization technique
    specifically designed for recurrent neural networks (RNNs). It generates a
    fixed dropout mask during the first call and applies it consistently to the
    hidden state across all time steps. This ensures the same units are dropped
    throughout the sequence, preventing co-adaptation of hidden units and
    improving generalization performance.

    Attributes:
      rate: The dropout probability (not the keep rate).
      rngs: An instance of nnx.Rngs.

    Example Usage:

    >>> from flax import nnx
    >>> import jax.numpy as jnp
    >>> rngs = nnx.Rngs(0)
    >>> dropout = RecurrentDropout(rngs=rngs, rate=0.2)
    >>> x = (jnp.ones((1, 10)), jnp.ones((1, 10)))  # Example hidden state and cell state
    >>> output = dropout(x)
    >>> print(output[0].shape)  # Output shape remains the same
    (1, 10)
    """
    def __init__(self, *, rngs: nnx.Rngs, rate: float):
        self.rate = rate
        self.rngs = rngs

    def __call__(self, x):
        """Applies a fixed dropout mask to the input.

        Args:
            x: the input to which the dropout mask will be applied.
        Returns:
            The input with the dropout mask applied.
        """
        if not hasattr(self, 'mask'):
            key = self.rngs.dropout()
            state = nnx.state(self).flat_state()
            state['mask'] = nnx.VariableState(
                type=nnx.Param, value=jax.random.bernoulli(key, 1 - self.rate, x[0].shape))
        h, c = x
        return h * state['mask'].value, c


class LSTMWithRecurrentDropout(nnx.OptimizedLSTMCell):
    """LSTM cell with recurrent dropout.

    This class extends the nnx.OptimizedLSTMCell by adding recurrent dropout,
    a regularization technique that applies dropout to the hidden state of the
    LSTM at each time step. This helps to prevent overfitting and improve
    generalization performance, particularly in tasks involving sequential data.

    Attributes:
        rngs: An instance of nnx.Rngs.
        in_features: The number of input features.
        hidden_features: The number of features in the hidden state of the LSTM.
        dropout_rate: The dropout probability for the recurrent connections.

    Example usage:

    >>> from flax import nnx
    >>> import jax.numpy as jnp
    >>> rngs = nnx.Rngs(0)
    >>> lstm = LSTMWithRecurrentDropout(rngs=rngs, in_features=10, hidden_features=20, dropout_rate=0.2)
    >>> h = jnp.ones((1, 20))  # (batch_size, hidden_features)
    >>> c = jnp.ones((1, 20))  # (batch_size, hidden_features)
    >>> x = jnp.ones((1, 10))  # (batch_size, in_features)
    >>> new_h, new_c = lstm((h, c), x, rngs=rngs)
    >>> print(new_h.shape, new_c.shape)
    (1, 20) (1, 20)
    """
    def __init__(self, *, rngs: nnx.Rngs, in_features: int, hidden_features: int, dropout_rate: float, **kwargs):
        super().__init__(in_features=in_features, hidden_features=hidden_features, rngs=rngs, **kwargs)
        self.recurrent_dropout = RecurrentDropout(rate=dropout_rate, rngs=rngs)

    def __call__(self, carry, x):
        h, c = carry
        new_h, new_c = super().__call__((h, c), x)
        new_h = self.recurrent_dropout(new_h)
        return new_h, new_c


class RNNWithRecurrentDropout(nnx.Module):
    """Recurrent Neural Network (RNN) with recurrent dropout.

    This module implements an RNN with an LSTM cell and recurrent dropout for
    processing sequential data. It applies dropout to the hidden state of the
    LSTM at each time step, helping to prevent overfitting and improve
    generalization.

    Attributes:
        rngs: An instance of nnx.Rngs.
        in_features: The number of input features.
        hidden_features: The number of features in the hidden state of the LSTM.
        dropout_rate: The dropout probability for the output layer.
        recurrent_dropout_rate: The dropout probability for the recurrent
                                connections in the LSTM.

    Example usage:

    >>> from flax import nnx
    >>> import jax.numpy as jnp
    >>> rngs = nnx.Rngs(0)
    >>> model = RNNWithRecurrentDropout(rngs=rngs, in_features=10, hidden_features=20)
    >>> x = jnp.ones((1, 3, 10))  # (batch_size, sequence_length, in_features)
    >>> output = model(x)
    >>> print(output.shape)
    (1, 1)
    """
    def __init__(self, *, rngs: nnx.Rngs, in_features: int,
                 hidden_features: int = 32, dropout_rate: float = 0.5, recurrent_dropout_rate: float = 0.25):
        cell = LSTMWithRecurrentDropout(in_features=in_features, hidden_features=hidden_features,
                                        rngs=rngs, dropout_rate=recurrent_dropout_rate)
        self.lstm = nnx.RNN(cell)
        self.dropout = nnx.Dropout(dropout_rate, rngs=rngs)
        self.dense = nnx.Linear(in_features=hidden_features, out_features=1, rngs=rngs)

    def __call__(self, x):
        x = self.lstm(x)
        x = self.dropout(x)
        x = x[:, -1, :]  # Use only the final hidden state
        return self.dense(x)
 ```
 
 I am not sure it works. I get a validation MAE of approx. 2.62 degrees while the Keras implementation achieved a validation MAE as low as 2.36 degrees. I guess that my implementation of the recurrent dropout is not correct.

**Comment by cgarciae** [comment, marked as answer]
Hey @sgkouzias, you can implement recurrent dropout by selecting a `rng_collection` different to `'dropout'` in `Dropout` constructor e.g. `'recurrent_dropout'`, and then broadcasting state for that RNG stream during `scan` (this is done internally by RNN). However, current RNN API is missing some options so I've created #4407 to address this. Here's a demo of the working solution:

```python

class LSTMWithRecurrentDropout(nnx.OptimizedLSTMCell):
  def __init__(
    self,
    *,
    rngs: nnx.Rngs,
    in_features: int,
    hidden_features: int,
    dropout_rate: float,
    **kwargs,
  ):
    super().__init__(
      in_features=in_features,
      hidden_features=hidden_features,
      rngs=rngs,
      **kwargs,
    )
    self.recurrent_dropout = nnx.Dropout(
      rate=dropout_rate, rng_collection='recurrent_dropout', rngs=rngs
    )

  def __call__(self, carry, x):
    h, c = carry
    new_h, new_c = super().__call__((h, c), x)
    new_h = jax.tree.map(self.recurrent_dropout, new_h)
    return new_h, new_c


class RNNWithRecurrentDropout(nnx.Module):
  def __init__(
    self,
    *,
    rngs: nnx.Rngs,
    in_features: int,
    hidden_features: int = 32,
    dropout_rate: float = 0.5,
    recurrent_dropout_rate: float = 0.25,
  ):
    cell = LSTMWithRecurrentDropout(
      in_features=in_features,
      hidden_features=hidden_features,
      rngs=rngs,
      dropout_rate=recurrent_dropout_rate,
    )
    self.lstm = nnx.RNN(cell, broadcast_rngs='recurrent_dropout')
    self.dropout = nnx.Dropout(dropout_rate, rngs=rngs)
    self.dense = nnx.Linear(
      in_features=hidden_features, out_features=1, rngs=rngs
    )

  def __call__(self, x):
    x = self.lstm(x)
    x = self.dropout(x)
    x = x[:, -1, :]  # Use only the final hidden state
    return self.dense(x)


model = RNNWithRecurrentDropout(
  in_features=32,
  hidden_features=64,
  dropout_rate=0.2,
  recurrent_dropout_rate=0.1,
  rngs=nnx.Rngs(0, recurrent_dropout=1),
)

x = jnp.ones((8, 10, 32))
y = model(x)

print(y.shape)
```
To test it locally you can install `flax` from the PR branch like this:
```bash
pip install git+https://github.com/google/flax@rnn-broadcast-rngs
```

**Reply by sgkouzias** [reply]
Hi @cgarciae,

Amazing. I tested the solution locally. I do confirm that the solution works. However, I get slightly better results. I get a validation MAE of approx. 2.61 degrees while the Keras implementation achieved a validation MAE as low as 2.27 degrees. I will check the entirity of my code in detail because it does not make sense. The trainining proccess is definitely faster! Thank you so much for your help.

**Reply by cgarciae** [reply]
@sgkouzias updated some details regarding RNG management which de-correlates RNGs a bit but I'm not sure this is enough to explain the difference. If you want to fully reproduce some results it might be best to reimplement the entire cell.

**Reply by sgkouzias** [reply]
By re-running the experiment got no better results. I will follow your advice and try to reimplement the entire `LSTMCell`. Thanks again for your valuable help and feedback.

![Figure_1](https://github.com/user-attachments/assets/2837cba2-46c8-4666-9704-6fd984944580)

**Reply by sgkouzias** [reply]
@cgarciae I updated the implementation as suggested. Used `glorot_uniform` as the default kernel initializer instead of `lecun_normal` and used bias term initialized with ones instead of zeros for the forget gate in order to comply with the Keras implementation ([documentation](https://github.com/mlresearch/v37/blob/gh-pages/jozefowicz15.pdf)). Interstingly and despite the notable and absolutely disappointing divergence between the results on the validation dataset (validation MAE of 2.61 degrees vs 2.29 degrees with the Keras implementation) it seems that the model implemented in Flax had better performance versus the Keras implementation in terms of test MAE! The test MAE score with the aformentioned re-implementation was ultimately 2.42 degrees versus test MAE of 2.48 degrees with the Keras implementation. I think that `recurrent_dropout` works :100:. Perhaps it would be nice for the user to choose whether the RNN will return sequences or not as the Keras framework provides this useful option.

For example like this:

```bash
class RNN(Module):
    """The ``RNN`` module takes any :class:`RNNCellBase` instance and applies it over a sequence
    using :func:`flax.nnx.scan`.
    """
    state_axes: Mapping[str, int | type[Carry] | None]

    def __init__(
            self,
            cell: nnx.RNNCellBase,
            time_major: bool = False,
            return_carry: bool = False,
            reverse: bool = False,
            keep_order: bool = False,
            unroll: int = 1,
            rngs: rnglib.Rngs | None = None,
            state_axes: Mapping[str, int | type[Carry] | None] | None = None,
            broadcast_rngs: filterlib.Filter = None,
            return_sequences: bool = False  # option to return sequences or not
    ):
        self.cell = cell
        self.time_major = time_major
        self.return_carry = return_carry
        self.reverse = reverse
        self.keep_order = keep_order
        self.unroll = unroll
        if rngs is None:
          rngs = rnglib.Rngs(0)
        self.rngs = rngs
        self.state_axes = state_axes or {...: Carry}  # type: ignore
        self.broadcast_rngs = broadcast_rngs
        self.return_sequences = return_sequences

    def __call__(
            self,
            inputs: Array,
            *,
            initial_carry: Carry | None = None,
            seq_lengths: Array | None = None,
            return_carry: bool | None = None,
            time_major: bool | None = None,
            reverse: bool | None = None,
            keep_order: bool | None = None,
            rngs: rnglib.Rngs | None = None,
    ):
        if return_carry is None:
            return_carry = self.return_carry
        if time_major is None:
            time_major = self.time_major
        if reverse is None:
            reverse = self.reverse
        if keep_order is None:
            keep_order = self.keep_order

        # Infer the number of batch dimensions from the input shape.
        # Cells like ConvLSTM have additional spatial dimensions.
        time_axis = 0 if time_major else inputs.ndim - (self.cell.num_feature_axes + 1)

        # make time_axis positive
        if time_axis < 0:
            time_axis += inputs.ndim

        if time_major:
            # we add +1 because we moved the time axis to the front
            batch_dims = inputs.shape[1 : -self.cell.num_feature_axes]
        else:
            batch_dims = inputs.shape[:time_axis]

        # maybe reverse the sequence
        if reverse:
            inputs = jax.tree_util.tree_map(
                lambda x: flip_sequences(
                    x,
                    seq_lengths,
                    num_batch_dims=len(batch_dims),
                    time_major=time_major,  # type: ignore
                ),
                inputs,
            )
        if rngs is None:
            rngs = self.rngs
        carry: Carry = (
            self.cell.initialize_carry(
                inputs.shape[:time_axis] + inputs.shape[time_axis + 1 :], rngs
            )
            if initial_carry is None
            else initial_carry
        )

        slice_carry = seq_lengths is not None and return_carry
        broadcast_rngs = nnx.All(nnx.RngState, self.broadcast_rngs)
        state_axes = StateAxes({broadcast_rngs: None, **self.state_axes})  # type: ignore

        # we use split_rngs with splits=1 and squeeze=True to get unique rngs
        # every time RNN is called
        @nnx.split_rngs(splits=1, only=self.broadcast_rngs, squeeze=True)
        @nnx.scan(
            in_axes=(state_axes, Carry, time_axis),
            out_axes=(Carry, (0, time_axis)) if slice_carry else (Carry, time_axis),
            unroll=self.unroll,
        )
        def scan_fn(
                cell: nnx.RNNCellBase, carry: Carry, x: Array
        ) -> tuple[Carry, Array] | tuple[Carry, tuple[Carry, Array]]:
            carry, y = cell(carry, x)
            if slice_carry:
                return carry, (carry, y)
            return carry, y

        scan_output = scan_fn(self.cell, carry, inputs)

        # Next we select the final carry. If a segmentation mask was provided and
        # return_carry is True we slice the carry history and select the last valid
        # carry for each sequence. Otherwise we just use the last carry.
        if slice_carry:
            assert seq_lengths is not None
            _, (carries, outputs) = scan_output
            # seq_lengths[None] expands the shape of the mask to match the
            # number of dimensions of the carry.
            carry = _select_last_carry(carries, seq_lengths)
        else:
            carry, outputs = scan_output

        if reverse and keep_order:
            outputs = jax.tree_util.tree_map(
                lambda x: flip_sequences(
                    x,
                    seq_lengths,
                    num_batch_dims=len(batch_dims),
                    time_major=time_major,  # type: ignore
                  ),
                outputs,
            )

        if not self.return_sequences:
            # Extract the final output for each sequence
            outputs = outputs[..., -1, :] if time_major else outputs[:, -1, :]
        
        if return_carry:
            return carry, outputs
        else:
            return outputs
```

**Reply by cgarciae** [reply]
Awesome!
